zebra-state/src/service/finalized_state/disk_db.rs

```diff
@@ -940,6 +943,111 @@ impl DiskDb {
 
     // Private methods
 
+    /// Tries to reuse an existing db after a major upgrade.
+    ///
+    /// If the current db version belongs to `restorable_db_versions`, the function moves a previous
+    /// db to a new path so it can be used again. It does so by merely trying to rename the path
+    /// corresponding to the db version directly preceding the current version to the path that is
+    /// used by the current db. If successful, it also deletes the db version file.
+    pub(crate) fn try_reusing_previous_db_after_major_upgrade(
+        restorable_db_versions: &[u64],
+        format_version_in_code: &Version,
+        config: &Config,
+        db_kind: impl AsRef<str>,
+        network: &Network,
+    ) {
+        if let Some(&major_db_ver) = restorable_db_versions
+            .iter()
+            .find(|v| **v == format_version_in_code.major)
+        {
+            let db_kind = db_kind.as_ref();
+
+            let old_path = config.db_path(db_kind, major_db_ver - 1, network);
+            let new_path = config.db_path(db_kind, major_db_ver, network);
+
+            let old_path = match fs::canonicalize(&old_path) {
+                Ok(canonicalized_old_path) => canonicalized_old_path,
+                Err(e) => {
+                    warn!("could not canonicalize {old_path:?}: {e}");
+                    return;
+                }
+            };
+
+            let cache_path = match fs::canonicalize(&config.cache_dir) {
+                Ok(canonicalized_cache_path) => canonicalized_cache_path,
+                Err(e) => {
+                    warn!("could not canonicalize {:?}: {e}", config.cache_dir);
+                    return;
+                }
+            };
+
+            // # Correctness
+            //
+            // Check that the path we're about to move is inside the cache directory.
+            //
+            // If the user has symlinked the state directory to a non-cache directory, we don't want
+            // to move it, because it might contain other files.
+            //
+            // We don't attempt to guard against malicious symlinks created by attackers
+            // (TOCTOU attacks). Zebra should not be run with elevated privileges.
+            if !old_path.starts_with(&cache_path) {
+                info!("skipped reusing previous state cache: state is outside cache directory");
+                return;
+            }
+
+            let opts = DiskDb::options();
+            let old_db_exists = DB::list_cf(&opts, &old_path).is_ok_and(|cf| !cf.is_empty());
+            let new_db_exists = DB::list_cf(&opts, &new_path).is_ok_and(|cf| !cf.is_empty());
+
+            if old_db_exists && !new_db_exists {
+                match fs::create_dir_all(&new_path) {
```

Comment:
```suggestion
                // Create the required folder path for the destination.
                // This is because we can't directly rename e.g. `state/v25/mainnet/` to `state/v26/mainnet/`
                // with `fs::rename()` because `state/v26/` does not exist.
                match fs::create_dir_all(&new_path) {
```